distributed/tests/test_active_memory_manager.py

```diff
@@ -903,6 +906,88 @@ async def test_RetireWorker_all_recipients_are_paused(c, s, a, b):
     assert await c.submit(inc, 1) == 2
 
 
+@gen_cluster(
+    client=True,
+    config={
+        "distributed.scheduler.active-memory-manager.start": True,
+        "distributed.scheduler.active-memory-manager.policies": [],
+    },
+    timeout=15,
+)
+async def test_RetireWorker_new_keys_arrive_after_all_keys_moved_away(
+    c, s: Scheduler, a: Worker, b: Worker
+):
+    """
+    If all keys have been moved off a worker, but then new keys arrive (due to task completion or `gather_dep`)
+    before the worker has actually closed, make sure we still retire it (instead of hanging forever).
+
+    This test is timing-sensitive. If it runs too slowly, it *should* `pytest.skip` itself.
+    However, this won't work until https://github.com/dask/distributed/issues/6239 is fixed.
+    So it may get stuck and fail at the 15s test timeout.
+
+    See https://github.com/dask/distributed/issues/6223 for motivation.
+    """
+    ws_a = s.workers[a.address]
+    ws_b = s.workers[b.address]
+    event = Event()
+
+    # Put 200 keys on the worker, so `_track_retire_worker` will sleep for 0.5s
+    # (Note that if you decrease the 200 to a very small value (say 2), this test will time out.
+    # But just because `extra` is restricted to only run on worker A, and A gets closed before `extra` can run,
+    # so not a deadlock (just user error)).
+    xs = c.map(lambda x: x, range(200), workers=[a.address])
+
+    while len(a.data) < len(xs):
+        await asyncio.sleep(0.1)
```

Comment:
```suggestion
    await wait(xs)
```
You need to wait for the worker to inform the scheduler